### PR TITLE
Adding prepend_docstring and append_docstring decorators

### PR DESCRIPTION
**Pull request by bsipocz**
This is to address #6757. Basically moving this functionality of astroquery upstream as both astropy and other packages could make use of it.

### CODE REVIEW

File: astropy/utils/decorators.py [hunk 1]
```diff
@@ -1101,3 +1100,87 @@ def set_docstring(obj):
...
+
+def prepend_docstring(doc, sections=None):
+    """
+    Decorator to prepend the given docstring to the decorated function's
+    after stripping out the list of sections provided.
+
+    Parameters
+    ----------
+    doc : str
+        Docstring to pretend.
```
Comment (pllim) [inline astropy/utils/decorators.py]: pretend -> prepend

File: astropy/utils/decorators.py [hunk 2]
```diff
@@ -1101,3 +1100,87 @@ def set_docstring(obj):
...
+
+    sections : list or None
+        Name of sections to remove from ``doc`` and replace with the
+        decorated doctring one. Default is None to remove no sections.
```
Comment (pllim) [inline astropy/utils/decorators.py]: doctring -> docstring

File: astropy/utils/decorators.py [hunk 3]
```diff
@@ -1101,3 +1100,87 @@ def set_docstring(obj):
...
+    def decor(func):
+        func.__doc__ = ("\n".join(_remove_sections(doc, sections)) +
+                        textwrap.dedent(func.__doc__))
+        return func
+    return decor
+
+
+def append_docstring(doc, sections=None):
+    """
+    Decorator to append the given docstring to the function's docstr after
```
Comment (pllim) [inline astropy/utils/decorators.py]: docstr -> docstring

File: astropy/utils/decorators.py [hunk 4]
```diff
@@ -1101,3 +1100,87 @@ def set_docstring(obj):
...
+    stripping out the list of sections provided.
+
+    Parameters
+    ----------
+    doc : str
+        Docstring to pretend.
```
Comment (pllim) [inline astropy/utils/decorators.py]: pretend -> append

File: astropy/utils/decorators.py [hunk 5]
```diff
@@ -1101,3 +1100,87 @@ def set_docstring(obj):
...
+
+    sections : list or None
+        Name of sections to remove from ``doc`` and replace with the
+        decorated doctring one. Default is None to remove no sections.
```
Comment (pllim) [inline astropy/utils/decorators.py]: doctring -> docstring

File: astropy/utils/decorators.py [hunk 6]
```diff
@@ -1101,3 +1100,87 @@ def set_docstring(obj):
...
+    """
+    def decor(func):
+        func.__doc__ = ("\n".join(_remove_sections(doc, sections)) +
```
Comment (pllim) [inline astropy/utils/decorators.py]: I wonder of `os.linesep` is more OS-agnostic than `"\n"`, although someone told me once that it doesn't matter. So, please ignore if you want.

File: astropy/utils/decorators.py [hunk 7]
```diff
@@ -1101,3 +1100,87 @@ def set_docstring(obj):
...
+    ``sections`` and dedent the whole thing.
+
+    Parameters
+    ----------
+    doc : str
+        Docstring to remove sections from.
+
+    sections : list or None
+        Name of sections to remove from ``doc`` and replace with the
+        decorated doctring one. Default is None to remove no sections.
```
Comment (pllim) [inline astropy/utils/decorators.py]: doctring -> docstring (although I realize that this docstring will never be rendered, but still...)

**Comment by pllim** [comment]
Also, does this warrant a "what's new" label?

File: astropy/utils/decorators.py (hunk 6 above)
Comment (bsipocz) [reply astropy/utils/decorators.py]: There are tons of `"\n"` in this file, so I would defer this for a possible future PR that deals with them consistently.

**Comment by bsipocz** [comment]
I don't think it needs a what's new (in my mind that's something very user facing, and this functionality is interesting only for devs), but maybe worth sending an e-mail to the astropy-dev list. @eteq?

**Comment by bsipocz** [comment]
@pllim - I'm not sure how to address your comment, the current examples are in the tests. But I've planned to look for examples in the codebase to start using this. That can either be part of this PR, or part of a follow-up.

**Comment by bsipocz** [comment]
Either one is fine but driven by the principle of only adding functionality that is being used, and given that this won't produce much conflicts in its current form, it's probably better to wait until I add at least a few modifications where this is being used.

**Comment by mhvk** [comment]
I think it would make sense to refer explicitly to `format_doc` in the docstrings here, as in a way these are special cases of that decorator (except that that one doesn't allow one to remove specific sections).

**Comment by bsipocz** [comment]
@mhvk -Oh, I didn't know about ``format_doc``. Do you think it would make sense to extend the functionality of that one with the option of removing sections? (Actually this was mostly triggered by numpydoc 0.7, that now raises warnings in cases of duplicated section names, previously it silently kept the last version of that section.)

**Comment by bsipocz** [comment]
(though I would argue the format_doc is not very descriptive of the functionality about combining docs)

**Comment by mhvk** [comment]
I wondered about combining them too, but then noticed how short the actual functions are. My 2¢ would be to just cross-reference. But cc @MSeifert04 (who wrote `format_doc`). 

p.s. Note that you actually touched a few lines, but probably as part of one of your major refactoring efforts, so I'm not surprised you didn't know about this. That does bring up the question, though, of how we can help other developers from knowing about these handy tricks...

**Comment by MSeifert04** [comment]
> I would argue the format_doc is not very descriptive of the functionality about combining docs

Yes, originally I thought the decorator would be really useful and wanted to add it in more places. But the problem was that I quickly realized that as developer I often look at the docstring of the function I'm working on and not having the complete docstring is a real drawback when developing. At least if you're too lazy to hunt down the remainder of the docstring. 

I even thought about removing the existing uses in nddata but I haven't been working on those for a while so I totally forgot about that.

> My 2¢ would be to just cross-reference.

Yes, that sounds like a good idea. Removing sections with `format_doc` would be nice-to-have but I'm not sure if that would be so easy because the `{__doc__}` placeholder could be anywhere, even inside a section.

**Comment by pllim** [comment]
So, is there consensus to refactor this into `format_doc`? Shall we remove the v3 milestone then?

**Comment by pllim** [comment]
Since today is supposed to be the :snowflake: , I am moving the milestone. FYI.

**Comment by bsipocz** [comment]
Oh, I can't believe the bot has caught up with me on this 😢

**Pull request closed without merging**